### PR TITLE
EFF-467 Port PartitionedSemaphore from effect v3

### PR DESCRIPTION
**Pull request by tim-smart**
## Summary
- Port `PartitionedSemaphore` into `packages/effect/src/PartitionedSemaphore.ts`, including v4 callback-based waiter registration and round-robin permit release across partitions.
- Add coverage in `packages/effect/test/PartitionedSemaphore.test.ts` for fairness, interruption safety, partial acquisition, and edge cases.
- Regenerate the effect barrel export (`packages/effect/src/index.ts`) via `pnpm codegen`.

## Validation
- `pnpm lint-fix`
- `pnpm test packages/effect/test/PartitionedSemaphore.test.ts`
- `pnpm check`
- `pnpm build`
- `pnpm docgen`

### CODE REVIEW

File: packages/effect/src/PartitionedSemaphore.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
+/**
+ * @since 4.0.0
```
Comment (tim-smart) [inline packages/effect/src/PartitionedSemaphore.ts]: Use the same since tags as effect v3
